File: Set (Dynamic Array)/Set.java
class mySet {
    private int size = 0;
    private String[] arr = new String[1];
    public int getIndex(String val) {
        for (int i = 0; i < arr.length; i++) {
            if (val.equals(arr[i])) return i; 
        }
        return -1;
    }
    public int getSize() {
        return size;
    }
    public void resize() {
        if (size >= arr.length){
            String[] temp = new String[arr.length*2];
            for (int i = 0; i < arr.length; i++){
                temp[i] = arr[i];
            }
            arr = temp;
        }
    }
    public boolean add(String val){
        if(contains(val)) return false;
        resize();
        arr[size++] = val;
        return true;
    }
    public boolean contains(String val) {
        if (getIndex(val) == -1) return false;
        return true;
    }
    public boolean remove(String val) {
        int index = getIndex(val);
        if(index == -1) return false;
        for (int i = index; i < size; i++) {
            arr[i] = arr[i+1];
        }
        return true;
    }
    public void print() {
        System.out.print("[");
        for (int i = 0; i < size; i++) {
            System.out.print(arr[i]);
            if (i != size - 1) System.out.print(", ");
        }
        System.out.println("]");
    }
}

public class Set {
    public static void main(String[] args) {
        mySet set = new mySet();
        set.add("Ahmad");
        set.add("Sultani");
        set.add("Dayanullah");
        set.print();
    }
}

File: Data Structures/cpp/Set (Dynamic Array)/set.cpp
#include <iostream>
using namespace std;

class set {
private:
    int size = 0;
    string* arr = new string[1*sizeof(arr)];
    int index(string val) {
        for (int i = 0; i <= size; i++) {
            if (val == arr[i]) return i;
        }
        return -1;
    }
    void resize() {
        if (size >= arr->size()){
            string* temp = new string[(arr->size()+1)*2*sizeof(arr)];
            for (int i = 0; i < arr->size(); i++) {
                temp[i] = arr[i];
            }
            arr = temp;
        }
    }
public:
    bool add(string val) {
        if (index(val) != -1) return false;
        resize();
        arr[size] = val;
        size++;
        return true;
    }
    int getSize() {
        return size;
    }
    bool remove(string val) {
        int idx = index(val);
        if (idx == -1) return false;
        for (int i = idx; i < size; i++) {
            arr[i] = arr[i+1];
        }
        size--;
        return true;
    }
    
};


int main() {
    ios::sync_with_stdio(0); cin.tie(0);
    set testing;
    cout << testing.getSize() << endl;
    cout << testing.add("Sultani") << endl;
    cout << "size: " << testing.getSize() << endl;
    cout << testing.add("XY") << endl;
    cout << "size: " << testing.getSize() << endl;
    cout << testing.add("XY") << endl;
    cout << testing.remove("XY") << endl;
    cout << "size: " << testing.getSize() << endl;
    cout << testing.remove("Ahmad") << endl;
    cout << testing.add("XY") << endl;
    cout << "size: " << testing.getSize() << endl;
    cout << testing.add("XY") << endl;
    for (int i = 0; i < 20; i++) {
        cout << testing.add("hi"+to_string(i)) << " ";
    }   
    cout << "size: " << testing.getSize() << endl;
    cout << testing.add("HALO");
    cout << "\n";
    return 0;
}